nemo_aligner/data/nlp/builders.py

```diff
@@ -321,29 +321,39 @@ def build_dataloader(
     pad_samples_to_global_batch_size=False,
     collate_fn=None,
     load_gbs=True,
+    use_random_sampler=True,
 ):
     """Buld dataloader given an input dataset."""
 
+    def create_batch_sampler(cls, **kwargs):
+        """Helper function to create a batch sampler."""
+        return cls(**kwargs)
+
     logging.info(f"Building dataloader with consumed samples: {consumed_samples}")
+    # Common parameters for batch sampler creation
+    common_params = {
+        "total_samples": len(dataset),
+        "consumed_samples": consumed_samples,
+        "micro_batch_size": mbs,
+        "data_parallel_rank": parallel_state.get_data_parallel_rank(),
+        "data_parallel_size": parallel_state.get_data_parallel_world_size(),
+        "drop_last": drop_last,
+        "global_batch_size": gbs,
+        "pad_samples_to_global_batch_size": pad_samples_to_global_batch_size,
+    }
+
     # Megatron sampler
-    if hasattr(cfg.model.data, "dataloader_type") and cfg.model.data.dataloader_type is not None:
-        if cfg.model.data.dataloader_type == "single":
+    if hasattr(cfg.model.data, "dataloader_type") and cfg.model.data.dataloader_type == "single":
+        if use_random_sampler:
             cls = MegatronPretrainingRandomBatchSampler if load_gbs else MegatronPretrainingRandomSampler
-            batch_sampler = cls(
-                total_samples=len(dataset),
-                consumed_samples=consumed_samples,
-                micro_batch_size=mbs,
-                data_parallel_rank=parallel_state.get_data_parallel_rank(),
-                data_parallel_size=parallel_state.get_data_parallel_world_size(),
-                drop_last=drop_last,
-                global_batch_size=gbs,
-                pad_samples_to_global_batch_size=pad_samples_to_global_batch_size,
-                seed=cfg.model.seed,
-            )
         else:
-            raise ValueError('cfg.data.dataloader_type must be "single"')
+            cls = MegatronPretrainingSampler
```

Comment:
I actually believe we should shuffle the validation set by default. This is safer when using `limit_val_batches` (to ensure the sample is representative of the full dataset), more consistent with the fact that the training set is also shuffled by default, and I believe we already released the current implementation so changing it now would be another breaking change.

This would require a different implementation though.

Reply:
Currently with `main`, PPO runs validation on a single and different batch on every validation run (as we haven't come around to fixing #27 yet). It makes no sense. 

While I also agree we should shuffle the validation set by default, I do think it's better keep it for a different issue/PR together with the accompanying expected behavior/implementation discussion, and maybe do that after we also fix #27. I wouldn't want to delay this issue being fixed, as currently validation is broken for PPO.

Reply:
Ah, I may be missing something here. Since `limit_val_batches` isn't supported in PPO I thought we'd always run validation on the full validation set and thus it shouldn't actually matter... what's actually happening?

(on my side I've just been using a small -- single-batch -- validation set so I didn't run into this issue)

Reply:
> what's actually happening?

Ah, I just talked to @gshennvm who reminded me of `num_val_samples` which I had forgotten about. I understand what's going on now.

I'm good with doing this (the shuffling of the validation set) in a subsequent PR then. @gshennvm and I agreed on how to implement it and I should be able to take care of it next week.